File: Example4OC/Example4OC/Classes/Map/BVRoomController.m
//
//  BVRoomController.m
//  LibosExample
//
//  Created by 黑眸智能 on 2021/11/2.
//

#import "BVRoomController.h"

@interface BVRoomController ()<BVSweeperMapViewRoomEditDelegate>

@property(nonatomic, weak) UIButton *nameBtn;
@property(nonatomic, weak) UIButton *fengeBtn;
@property(nonatomic, weak) UIButton *hebingBtn;

@end

@implementation BVRoomController

- (void)viewDidLoad {
    [super viewDidLoad];
}

- (NSString *)getNavTitle {
    return @"房间编辑";
}

- (BVSweeperMapViewMode)getMapMode {
    return  BVSweeperMapViewModeRoomEditMode;
}

- (void)initView {
    [super initView];
    
    self.mapView.editRoomDelegate = self;
    
    UIButton *recoverBtn = [UIButton buttonWith:@"恢复选择"];
    [self.view addSubview:recoverBtn];
    [recoverBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self.view);
        make.top.equalTo(self.logView.mas_bottom).offset(10);
    }];
    [recoverBtn addTarget:self action:@selector(recover) forControlEvents:UIControlEventTouchUpInside];
    
    UIButton *nameBtn = [self createButton:@"重命名"];
    [self.view addSubview:nameBtn];
    [nameBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(recoverBtn.mas_bottom).offset(10);
        make.left.equalTo(self.view).offset(10);
    }];
    [nameBtn addTarget:self action:@selector(name) forControlEvents:UIControlEventTouchUpInside];
    self.nameBtn = nameBtn;
    
    UIButton *fengeBtn = [self createButton:@"分割"];
    [self.view addSubview:fengeBtn];
    [fengeBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(recoverBtn.mas_bottom).offset(10);
        make.centerX.equalTo(self.view);
    }];
    [fengeBtn addTarget:self action:@selector(fenge) forControlEvents:UIControlEventTouchUpInside];
    self.fengeBtn = fengeBtn;
    
    UIButton *hebingBtn = [self createButton:@"合并"];
    [self.view addSubview:hebingBtn];
    [hebingBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(recoverBtn.mas_bottom).offset(10);
        make.right.equalTo(self.view).offset(-10);
    }];
    [hebingBtn addTarget:self action:@selector(hebing) forControlEvents:UIControlEventTouchUpInside];
    self.hebingBtn = hebingBtn;
    
    UIButton *fengeOKBtn = [self createButton:@"获取分割点"];
    [self.view addSubview:fengeOKBtn];
    [fengeOKBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(hebingBtn.mas_bottom).offset(10);
        make.left.equalTo(self.view).offset(10);
    }];
    [fengeOKBtn addTarget:self action:@selector(fengeOk) forControlEvents:UIControlEventTouchUpInside];
    
    UIButton *getRoomInfoBtn = [self createButton:@"获取房间属性"];
    [self.view addSubview:getRoomInfoBtn];
    [getRoomInfoBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(fengeOKBtn.mas_bottom).offset(10);
        make.left.equalTo(self.view).offset(10);
    }];
    [getRoomInfoBtn addTarget:self action:@selector(getRoomInfo) forControlEvents:UIControlEventTouchUpInside];
    
    UIButton *adjustBtn = [self createButton:@"设置房间属性"];
    [self.view addSubview:adjustBtn];
    [adjustBtn mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerY.equalTo(getRoomInfoBtn);
        make.left.equalTo(getRoomInfoBtn.mas_right).offset(10);
    }];
    [adjustBtn addTarget:self action:@selector(tryAdjustRoom) forControlEvents:UIControlEventTouchUpInside];
    
    [self log:@"选择房间，再选择操作"];
    [self log:@"分割房间之后再点击<获取分割点>按钮"];
    
    [self recover];
}

-(UIButton *)createButton:(NSString *)title {
    UIButton *button = [UIButton buttonWith:title];
    [button setTitleColor:UIColor.systemGrayColor forState:UIControlStateDisabled];
    return button;
}

-(void)recover {
    [self.mapView resetRoomCheck];
    
    self.nameBtn.enabled = NO;
    self.fengeBtn.enabled = NO;
    self.hebingBtn.enabled = NO;
}

-(void)name {
    BVDrawMapRoom *room = [self.mapView getCheckRooms].firstObject;
    if(room != nil) {
        NSString *oldName = room.roomName;
        NSInteger no = room.no;
        
        [self log:[NSString stringWithFormat:@"将房间号： %ld 原名称：%@ 改名", no, oldName]];
    }
    
    [self recover];
}

-(void)fenge {
    if(![self.mapView canSliptRoom]) {
        [self log:@"房间数量到达上限"];
        [BVHUD showInfo:@"房间数量到达上限"];
        [self recover];
        return;
    }
    
    [self.mapView changeState:BVSweeperMapViewStateRoomSplitState];
    
    [self log:@"分割房间  可以在屏幕上分割)"];
}

-(void)fengeOk {
    NSNumber *roomId = [self.mapView getCheckRoomIds].firstObject;
    
    CGPoint p1 = [self.mapView getRoomSplitStartPoint];
    CGPoint p2 = [self.mapView getRoomSplitEndPoint];
    
    [self log:[NSString stringWithFormat:@"分割房间  房间号：%@  分割点 起点：%@  重点：%@", roomId.stringValue, NSStringFromCGPoint(p1), NSStringFromCGPoint(p2)]];
    
    [self recover];
}

-(void)hebing {
    NSArray<NSNumber *> *roomIds = [self.mapView getCheckRoomIds];
    
    [self log:[NSString stringWithFormat:@"合并房间  房间号：%@", roomIds]];
    
    [self recover];
}

-(void)getRoomInfo {
    NSArray<BVDrawMapRoom *> * rooms = [self.mapView getAllRoomInfo];
    NSMutableString *str = [NSMutableString string];
    [str appendString:@"当前房间属性：\n"];
    for (BVDrawMapRoom *item in rooms) {
        [str appendFormat:@"房间名称：%@ %@\n", item.roomName, item.roomProperty];
    }
    [self log:str];
}

-(void)tryAdjustRoom {
    NSInteger no = [self.mapView getCheckRoomIds].firstObject.integerValue;
    [self.mapView adjustRoomPropertyWith:no CleanTimes:nil FanLevel:[NSNumber numberWithInt:2] WaterLevel:nil];
    NSArray<BVDrawMapRoom *> * rooms = [self.mapView adjustRoomPropertyWith:no CleanTimes:[NSNumber numberWithInt:2] FanLevel:nil WaterLevel:nil];
    
    NSMutableString *str = [NSMutableString string];
    [str appendString:@"调整后的房间属性：\n"];
    for (BVDrawMapRoom *item in rooms) {
        [str appendFormat:@"房间名称：%@ %@\n", item.roomName, item.roomProperty];
    }
    [self log:str];
}

- (void)mapView:(BVSweeperMapView *)view didCheckRoomCount:(NSInteger)count {
    if(count == 0) {
        self.nameBtn.enabled = NO;
        self.fengeBtn.enabled = NO;
        self.hebingBtn.enabled = NO;
    } else if (count == 1) {
        self.nameBtn.enabled = YES;
        self.fengeBtn.enabled = YES;
        self.hebingBtn.enabled = NO;
    } else if (count == 2) {
        self.nameBtn.enabled = NO;
        self.fengeBtn.enabled = NO;
        self.hebingBtn.enabled = YES;
    }
}

@end

File: Example4Swift/Example4Swift/Classes/Map/BVRoomController.swift
//
//  BVRoomController.swift
//  Libos
//
//  Created by 黑眸智能 on 2021/11/1.
//

import UIKit
import BVSweeperKit

class BVRoomController: BVMapController {
    
    var nameBtn: UIButton!
    var fenggeBtn: UIButton!
    var hebingBtn: UIButton!

    override func viewDidLoad() {
        super.viewDidLoad()
    }
}

extension BVRoomController {
    override func getNavTitle() -> String {
        "房间编辑"
    }
    
    override func getMapMode() -> BVSweeperMapViewMode {
        .roomEditMode
    }
    
    override func initView() {
        super.initView()
        
        self.mapView.editRoomDelegate = self
        
        
        let recoverBtn = UIButton.buttonWith(title: "恢复选择")
        self.view.addSubview(recoverBtn)
        recoverBtn.snp.makeConstraints { make in
            make.centerX.equalToSuperview()
            make.top.equalTo(logView.snp.bottom).offset(10)
        }
        recoverBtn.addTarget(self, action: #selector(recover), for: .touchUpInside)
        
        nameBtn = self.createButton(title: "重命名")
        self.view.addSubview(nameBtn)
        nameBtn.snp.makeConstraints { make in
            make.top.equalTo(recoverBtn.snp.bottom).offset(10)
            make.left.equalToSuperview().offset(10)
        }
        nameBtn.addTarget(self, action: #selector(name), for: .touchUpInside)
        
        fenggeBtn = self.createButton(title: "分割")
        self.view.addSubview(fenggeBtn)
        fenggeBtn.snp.makeConstraints { make in
            make.top.equalTo(recoverBtn.snp.bottom).offset(10)
            make.centerX.equalToSuperview()
        }
        fenggeBtn.addTarget(self, action: #selector(fenge), for: .touchUpInside)
        
        hebingBtn = self.createButton(title: "合并")
        self.view.addSubview(hebingBtn)
        hebingBtn.snp.makeConstraints { make in
            make.top.equalTo(recoverBtn.snp.bottom).offset(10)
            make.right.equalToSuperview().offset(-10)
        }
        hebingBtn.addTarget(self, action: #selector(hebing), for: .touchUpInside)
        
        
        let fengeOKBtn = self.createButton(title: "获取分割点")
        self.view.addSubview(fengeOKBtn)
        fengeOKBtn.snp.makeConstraints { make in
            make.top.equalTo(hebingBtn.snp.bottom).offset(10)
            make.left.equalToSuperview().offset(10)
        }
        fengeOKBtn.addTarget(self, action: #selector(fengeOk), for: .touchUpInside)
        
        log("选择房间，再选择操作")
        log("分割房间之后再点击<获取分割点>按钮")
        
        recover()
    }
    
    func createButton(title: String) -> UIButton {
        let button = UIButton.buttonWith(title: title)
        button.setTitleColor(.systemGray, for: .disabled)
        return button
    }
}

@objc extension BVRoomController {
    func recover() {
        self.mapView.resetRoomCheck()
        self.nameBtn.isEnabled = false
        self.fenggeBtn.isEnabled = false
        self.hebingBtn.isEnabled = false
    }
    
    func name() {
        guard let model = self.mapView.getCheckRooms().first else {
            return
        }
        let oldName = model.roomName ?? "null"
        let no = model.no
        
        self.log("将房间号： \(no) 原名称：\(oldName) 改名")
        
        recover()
    }
    
    func fenge() {
        guard self.mapView.canSliptRoom() else {
            log("房间数量到达上限")
            BVHUD.showInfo("房间数量到达上限")
            recover()
            return
        }
        self.mapView.changeState(.roomSplitState)
        log("分割房间  可以在屏幕上分割)")
    }
    
    func fengeOk() {
        guard let roomId = self.mapView.getCheckRoomIds().first else {
            return
        }
        
        let p1 = self.mapView.getRoomSplitStartPoint()
        let p2 = self.mapView.getRoomSplitEndPoint()
        
        log("分割房间  房间号：\(roomId)  分割点 起点：\(p1)  重点：\(p2)")
        
        recover()
    }
    
    func hebing() {
        let roomIds = self.mapView.getCheckRoomIds()
        log("合并房间  房间号：\(roomIds)")
        recover()
    }
}

extension BVRoomController: BVSweeperMapViewRoomEditDelegate {
    func mapView(_ view: BVSweeperMapView, didCheckRoomCount count: Int) {
        if count == 0 {
            self.nameBtn.isEnabled = false
            self.fenggeBtn.isEnabled = false
            self.hebingBtn.isEnabled = false
        } else if count == 1 {
            self.nameBtn.isEnabled = true
            self.fenggeBtn.isEnabled = true
            self.hebingBtn.isEnabled = false
        } else if count == 2 {
            self.nameBtn.isEnabled = false
            self.fenggeBtn.isEnabled = false
            self.hebingBtn.isEnabled = true
        }
    }
}
